src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -419,16 +490,25 @@ protected FlowNode next(@Nonnull FlowNode current, @Nonnull Collection<FlowNode>
                 // Terminating a parallel scan
                 FlowNode temp = hitParallelStart();
                 if (temp != null) { // Start node for current parallel block now that it is done
+                    nextType = NodeType.PARALLEL_START;
                     return temp;
                 }
             } else if (!blackList.contains(p)) {
+                if (p instanceof BlockStartNode && p.getAction(ThreadNameAction.class) != null) {
+                    nextType = NodeType.PARALLEL_BRANCH_START;
+                } else if (ForkScanner.isParallelEnd(p)) {
+                    nextType = NodeType.PARALLEL_END;
+                } else {
+                    nextType = NodeType.NORMAL;
+                }
```

Comment:
It's a good point: it is a sort of state machine, but the transition rules are rather complex (of necessity, unfortunately).   The currentType/nextType *might* be usable to feed some of the internal logic -- it might simplify things, it might not.  

Currently the discovery of a parallelStartNode for the type fields is a bit hack though, so it's not as trustworthy yet for that purpose. 

May be worth revisiting this in the future if ForkScanner needs to be touched again after this.